[validation] format dates in validation messages and support validations on date type

vicbergquist opened the pull request:
<!-- Thank you for contributing to Sanity. Please read our [Code of Conduct](https://github.com/sanity-io/sanity/blob/next/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/sanity-io/sanity/blob/next/CONTRIBUTING.md) before submitting a PR.

To help us review your Pull Request, please make sure you follow the steps and guidelines below.

It's OK to open a Pull Request to start a discussion/ask for help, but it should then be created as a [Draft Pull Request](https://github.blog/2019-02-14-introducing-draft-pull-requests/). 

Do not delete the instructional comments. -->

**Type of change (check at least one)**

- [x]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Documentation (fix or update to documentation)
- [ ]  Maintenance
- [ ]  Other, please describe:

**Does this change require a documentation update? (Check one)**

- [ ]  Yes
- [x]  No

**Current behavior**

<!-- Reference/link the relevant issue and/or describe the behavior. -->
When you have validation on date fields, the date isn't formatted when showing in the validation message, which makes it hard to read.

<img width="421" alt="Screenshot 2020-05-26 at 16 46 13" src="https://user-images.githubusercontent.com/25737281/83037744-733bc400-a03c-11ea-81f5-2d8f05ef496d.png">

Also, validation hasn't been possible on the `date` type.

**Description**

<!-- Please include a summary of the changes this PR introduces and/or which issue is fixed. Please also include relevant motivation and context of why this PR is necessary. -->
This gets the options from the field to be validated, and uses it in the validator to format the dates using either the default format, or the format specified in options.

This adds the same validation methods to the `date` type. 



**Note for release**

<!-- Please include a high level summary of the changes this PR introduce. The intended audience is both editors and developers. If it's introducing a new feature, remember to link to docs/blogpost, if it's a bugfix, please describe the bug in non-technical terms (e.g. how a user/developer may have experienced it).
For inspiration, check out the release notes from an earlier release: https://github.com/sanity-io/sanity/releases/tag/v0.142.0 -->
- Fixed an issue where `date` wouldn't get the same built-in validation methods as `datetime`
- Fixed date formatting in validation messages

**Checklist** 

- [x]  I have read the [Contributing Guidelines](https://github.com/sanity-io/sanity/blob/next/CONTRIBUTING.md)
- [x]  The PR title is appropriately formatted: `[some-package] PR title (#123)`
- [x]  The code is linted
- [x]  The test suite is passing

saasen commented on packages/@sanity/validation/src/validateDocument.js:
```diff
@@ -131,7 +131,13 @@ function validatePrimitive(item, type, path, options) {
+        document: options.document,
+        path,
+        fieldOptions: options.type && options.type.options,
+        typeName: options.type && options.type.name
```
Since it's now two fields from the `type` we are sending in here, would it make more sense to send in the whole `type`? So we don't have to change this and add a new field if we happen to need another field from the `type`?

vicbergquist replied:
I guess so 🤔

saasen replied:
It would also align with the other fields passed in already (`document` and `parent`). 👍 I'll approve given these changes (if you agree, of course).

vicbergquist replied:
ok! Will change it

vicbergquist commented:
Also pushed a change to enable validations on the `date` type

vicbergquist commented:
I guess I also have to now add `moment.js` to the `package.json`? @rexxars

rexxars commented:
> I guess I also have to now add `moment.js` to the `package.json`? @rexxars

I'd prefer if we could use [date-fns](https://date-fns.org/v1.29.0/docs/format)@^1.29.0 - we're trying to migrate away from moment, and date-fns is already used by the form builder etc. And yes, it needs to be added to package.json

vicbergquist commented:
> > I guess I also have to now add `moment.js` to the `package.json`? @rexxars
> 
> I'd prefer if we could use [date-fns](https://date-fns.org/v1.29.0/docs/format)@^1.29.0 - we're trying to migrate away from moment, and date-fns is already used by the form builder etc. And yes, it needs to be added to package.json

@rexxars Ok! I've replaced it with `date-fns` and added it to the `package.json` 👍
